python/repositories.bzl

```diff
@@ -1,8 +1,164 @@
 """
 
+load("//python/private:toolchains_repo.bzl", "PLATFORMS", "toolchains_repo")
+load("//python/private:versions.bzl", "MINOR_MAPPING", "RELEASE_DATE", "RELEASE_URL", "TOOL_VERSIONS")
+
 def py_repositories():
     # buildifier: disable=print
     print("py_repositories is a no-op and is deprecated. You can remove this from your WORKSPACE file")
+
+########
+# Remaining content of the file is only used to support toolchains.
+########
+_DOC = "Fetch external tools needed for python toolchain."
+_ATTRS = {
+    "platform": attr.string(mandatory = True, values = PLATFORMS.keys()),
+    "python_version": attr.string(mandatory = True, values = TOOL_VERSIONS.keys() + MINOR_MAPPING.keys()),
+}
+
+def _python_repo_impl(rctx):
+    version = rctx.attr.python_version
+    if version in MINOR_MAPPING:
+        version = MINOR_MAPPING[version]
+    release_filename = "cpython-{version}-{platform}-pgo+lto-{release_date}.tar.zst".format(
+        release_date = RELEASE_DATE,
+        version = version,
+        platform = rctx.attr.platform,
+    )
+    url = "{release_url}/{release_filename}".format(
+        release_url = RELEASE_URL,
+        release_filename = release_filename,
+    )
+    rctx.download(
+        url = url,
+        integrity = TOOL_VERSIONS[version][rctx.attr.platform],
+        output = release_filename,
+    )
+    unzstd = rctx.which("unzstd")
+    if not unzstd:
+        url = rctx.attr._zstd_url.format(version = rctx.attr._zstd_version)
+        rctx.download_and_extract(
+            url = url,
+            sha256 = rctx.attr._zstd_sha256,
+        )
+        working_directory = "zstd-{version}".format(version = rctx.attr._zstd_version)
+        rctx.execute(
+            ["make", "--jobs=4"],
+            timeout = 600,
+            quiet = True,
+            working_directory = working_directory,
+        )
+        zstd = "{working_directory}/zstd".format(working_directory = working_directory)
+        unzstd = "./unzstd"
+        rctx.symlink(zstd, unzstd)
+
+    exec_result = rctx.execute([
+        "tar",
+        "--extract",
+        "--strip-components=2",
+        "--use-compress-program={unzstd}".format(unzstd = unzstd),
+        "--file={}".format(release_filename),
+    ])
+    if exec_result.return_code:
+        fail(exec_result.stderr)
+
+    build_content = """\
+# Generated by python/repositories.bzl
+
+load("@bazel_tools//tools/python:toolchain.bzl", "py_runtime_pair")
+
+filegroup(
+    name = "files",
+    srcs = glob(["bin/**", "lib/**", "include/**", "share/**"]),
+)
+
+exports_files(["bin/python3"])
+
+py_runtime(
+    name = "py2_runtime",
+    interpreter_path = "/bin/false",
+    python_version = "PY2",
+)
+
+py_runtime(
+    name = "py3_runtime",
+    interpreter = "bin/python3",
+    python_version = "PY3",
+)
+
+py_runtime_pair(
+    name = "python_runtimes",
+    py2_runtime = ":py2_runtime",
+    py3_runtime = ":py3_runtime",
+)
+"""
+    rctx.file("BUILD.bazel", build_content)
```

Comment:
Consider returning here the values of parameters to the bazel rule, which would make the rule reproducible. See https://docs.bazel.build/versions/main/skylark/repository_rules.html#implementation-function

For example:
```starlark
    shas = dict(**repository_ctx.attr.sha256)
    shas[arch] = sha256 # put in the sha256 that we get when we download the interpreter
    return {
        "name": repository_ctx.attr.name,
        "sha256": shas,
        "timestamp": repository_ctx.attr.timestamp,
        "version": repository_ctx.attr.version,
    }
```

This assumes that the repository rule accepts:
* `name` for the repository.
* `sha256` dict where we have a mapping from `platform` to `sha256` of the downloaded artifact.
* `timestamp` is the timestamp value in the download url
* `version` the python version to download.

For example, the rule can be executed with the following params:
```starlark
        name = "python_interpreter",
        sha256 = {
            "x86_64-apple": "12c5bce1b48d2b896049ec2524648d5429eba982539bc571d43c1a2ec3997630",
            "x86_64-linux": "33cb6a4895418b9de2b770f8ab72d1fd2dbebb95747a81f6d1824c1900062df8",
        },
        timestamp = "20211017T1616",
        version = "3.9.7",
```

If the sha256 is missing from the dict (e.g. someone wants to add a download for ARM apple), then `bazel` would output a warning on how to change the input parameters.

Reply:
Addressed in https://github.com/bazelbuild/rules_python/pull/618/commits/0d84ae6114368208704eaa26bf5b888247118221.